library/core/src/ptr/non_null.rs

```diff
@@ -430,6 +432,10 @@ impl<T: ?Sized> NonNull<T> {
     #[must_use = "this returns the result of the operation, \
                   without modifying the original"]
     #[inline]
+    //Address presaervation
+    #[ensures(|result: &NonNull<U>|  result.as_ptr() as *const () as usize == self.as_ptr() as *const () as usize )]
+    //Ensures size of pointer types is same and pointer is properly aligned for new type 'U'
+    #[ensures(|result: &NonNull<U>|  core::mem::size_of::<*mut U>() == core::mem::size_of::<*mut T>() || (self.as_ptr() as *const () as usize) % core::mem::align_of::<U>() == 0)]
```

Comment:
What is the purpose of this post-condition?

Reply:
This post condition ensured that either the size of pointer is the same or pointer is properly aligned for the new type. It was a little redundant to the first pre condition so I have updated it to::` #[ensures(|result: &NonNull<U>| (self.as_ptr() as *const () as usize) % core::mem::align_of::<U>() == 0)]`